### DISCUSSION TITLE
New diagram type for SQL schemas

### DISCUSSION THREAD

**Discussion by jaads** [mermaid-js/mermaid#5062] [Ideas]
As a developer I would love to have the possibility to create a visualisation of SQL schemas with mermaid. This would mainly mean to be able to create tables with the names of the columns and their data types as well as the column constraints to indicate primary and foreign keys etc. So actual what the ERD type is. But for a foreign key it would be very nice to have an arrow **to the targeting primary key of the foreign table**. 

As an example let's say an article can be written by only one author. This is ho it looks like currently:

```mermaid
erDiagram
  articles {
		int id PK
		int author_id	FK
	}
	authors {
		int id PK
	}
authors ||--o{ articles : "writes"
```
But the actual SQL database schema would be more accurately visualised if it would look like this, because cardinalities  doesn't technically exists in an SQL database, only foreign keys exist.   
<img width="492" alt="Screenshot 2023-11-24 at 14 59 19" src="https://github.com/mermaid-js/mermaid/assets/34009329/8d5b8ad0-897c-4cc0-ab8f-0cf2efd38233">

As an alternative, the ERD could be made database independent and treated as something only semantically at which point a completely new type for SQL which has primary and foreign keys included and the arrow I mentioned above. Then mermaid would have one semantic diagram and one technical diagram for visualising entity relations.

**Comment by lukeSmth** [comment]
Do you have an idea of how composite keys would be represented?

**Reply by JosephArenas-ACF** [reply]
Not necessarily for the new diagram type being suggested, but an enhancement to ERDs, it would be great if there was a way to specify composite keys.

For example, the component columns of composite unique or foreign keys could be specified as being together in the key by adding a numeric suffix like this:
  - first composite foreign key (col_1, col_2):
    - col_1 FK1
    - col_2 FK1
  - second composite foreign key (col_3, col_4):
    - col_3 FK2
    - col_4 FK2
  - first composite unique key (col_5, col_6):
    - col_5 UK1
    - col_6 UK1
  - second composite unique key (col_7, col_8):
    - col_7 UK2
    - col_8 UK2

Non-composite keys would just be indicated by FK or UK.
